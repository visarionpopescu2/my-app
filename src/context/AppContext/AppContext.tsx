import React from 'react';
import { APP_STATE } from "types";

const initialState: APP_STATE = {
  user: undefined,
  articles: [],
  setArticles: () => {}
}

const AppContext = React.createContext(initialState);

export default AppContext
