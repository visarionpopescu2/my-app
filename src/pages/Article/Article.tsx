import {ReactNode, useContext, useEffect, useState} from 'react';
import { Link, useParams } from 'react-router-dom';
import { ARTICLE } from 'types';
import { Breadcrumb, Typography } from 'antd';
import AppContext from 'context/AppContext/AppContext';
import { ROUTES } from 'routes';

function Article() {
  const { articles } = useContext(AppContext);
  const [article, setArticle] = useState<ARTICLE>();
  const params = useParams();
  
  useEffect(() => {
    if (params.id) {
      setArticle(articles.find((item) => item.id +'' === params.id));
    }
    
  }, [params, articles]);
  
  const breadcrumbItems: { title: string | ReactNode }[] = [
    { title: <Link to={ROUTES.Articles}>Articles</Link> },
    { title: <Typography.Text>Details</Typography.Text> },
  ];
  
  return (
    <div>
      <Breadcrumb items={breadcrumbItems}/>
      <Typography.Title level={3}>{article?.title}</Typography.Title>
      <Typography.Paragraph>{article?.body}</Typography.Paragraph>
    </div>
  );
};

export default Article;
