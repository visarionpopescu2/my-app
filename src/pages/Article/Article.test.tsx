import { render } from '@testing-library/react';
import Article from './Article';
import Router, { MemoryRouter } from 'react-router-dom';
import AppContext from 'context/AppContext/AppContext';

const ARTICLE_ID = 1;
const USER = {
  id: '1',
  name: 'name',
  username: 'username',
  email: 'email',
  phone: '123445',
  website: 'www.website.com',
  company: { name: 'company', bs: 'business', catchPhrase: 'description' },
  address: { street: 'street', suite: 'suite', zipcode: '1234', city: 'city', geo: { lat: '11', lng: '22' }}
};
const setArticles = jest.fn();
const ARTICLE = { userId: 1, id: ARTICLE_ID, title: 'title', body: 'body' };

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: jest.fn(),
}));

describe('Article', () => {
  beforeEach(() => {
    Object.defineProperty(window, 'matchMedia', {
      writable: true,
      value: jest.fn().mockImplementation(query => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // deprecated
        removeListener: jest.fn(), // deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });
  });
  
  it('should render Article page with article', () => {
    jest.spyOn(Router, 'useParams').mockReturnValue({ id: ARTICLE_ID + '' })
    let view = render(
      <MemoryRouter>
        <AppContext.Provider value={{ user: USER, articles: [ ARTICLE ], setArticles: setArticles }}>
          <Article />
        </AppContext.Provider>
      </MemoryRouter>
    );
    expect(view.asFragment()).toMatchSnapshot();
  });

  it('should render Article page without article', () => {
    jest.spyOn(Router, 'useParams').mockReturnValue({ id: 'random' })
    let view = render(
      <MemoryRouter>
        <AppContext.Provider value={{ user: USER, articles: [ ARTICLE ], setArticles: setArticles }}>
          <Article />
        </AppContext.Provider>
      </MemoryRouter>
    );
    expect(view.asFragment()).toMatchSnapshot();
  });
});
