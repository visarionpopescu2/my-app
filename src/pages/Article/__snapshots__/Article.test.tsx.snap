// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article should render Article page with article 1`] = `
<DocumentFragment>
  <div>
    <nav
      class="ant-breadcrumb css-dev-only-do-not-override-2i2tap"
    >
      <ol>
        <li>
          <span
            class="ant-breadcrumb-link"
          >
            <a
              href="/articles"
            >
              Articles
            </a>
          </span>
        </li>
        <li
          aria-hidden="true"
          class="ant-breadcrumb-separator"
        >
          /
        </li>
        <li>
          <span
            class="ant-breadcrumb-link"
          >
            <span
              class="ant-typography css-dev-only-do-not-override-2i2tap"
            >
              Details
            </span>
          </span>
        </li>
      </ol>
    </nav>
    <h3
      class="ant-typography css-dev-only-do-not-override-2i2tap"
    >
      title
    </h3>
    <div
      class="ant-typography css-dev-only-do-not-override-2i2tap"
    >
      body
    </div>
  </div>
</DocumentFragment>
`;

exports[`Article should render Article page without article 1`] = `
<DocumentFragment>
  <div>
    <nav
      class="ant-breadcrumb css-dev-only-do-not-override-2i2tap"
    >
      <ol>
        <li>
          <span
            class="ant-breadcrumb-link"
          >
            <a
              href="/articles"
            >
              Articles
            </a>
          </span>
        </li>
        <li
          aria-hidden="true"
          class="ant-breadcrumb-separator"
        >
          /
        </li>
        <li>
          <span
            class="ant-breadcrumb-link"
          >
            <span
              class="ant-typography css-dev-only-do-not-override-2i2tap"
            >
              Details
            </span>
          </span>
        </li>
      </ol>
    </nav>
    <h3
      class="ant-typography css-dev-only-do-not-override-2i2tap"
    />
    <div
      class="ant-typography css-dev-only-do-not-override-2i2tap"
    />
  </div>
</DocumentFragment>
`;
